lib/Sema/TypeCheckCodeCompletion.cpp

```diff
@@ -1153,31 +1203,45 @@ sawSolution(const constraints::Solution &S) {
 
   auto Key = std::make_pair(BaseTy, ReferencedDecl);
   auto Ret = BaseToSolutionIdx.insert({Key, Results.size()});
-  if (!Ret.second && ExpectedTy) {
-    Results[Ret.first->getSecond()].ExpectedTypes.push_back(ExpectedTy);
-  } else {
+  if (Ret.second) {
     bool ISDMT = S.isStaticallyDerivedMetatype(ParsedExpr);
-    bool SingleExprBody = false;
+    bool SingleExprBody = isSingleExpressionBodyCompletion(CS, CompletionExpr);
     bool DisallowVoid = ExpectedTy
                             ? !ExpectedTy->isVoid()
                             : !ParentExpr && CS.getContextualTypePurpose(
                                                  CompletionExpr) != CTP_Unused;
 
-    if (!ParentExpr) {
-      if (CS.getContextualTypePurpose(CompletionExpr) == CTP_ReturnSingleExpr)
-        SingleExprBody = true;
-    } else if (auto *ParentCE = dyn_cast<ClosureExpr>(ParentExpr)) {
-      if (ParentCE->hasSingleExpressionBody() &&
-          ParentCE->getSingleExpressionBody() == CompletionExpr) {
-        ASTNode Last = ParentCE->getBody()->getLastElement();
-        if (!Last.isStmt(StmtKind::Return) || Last.isImplicit())
-          SingleExprBody = true;
```

Comment:
Could you remind me why this check is needed (why `return (ParentCE->hasSingleExpressionBody() &&
           ParentCE->getSingleExpressionBody() == CompletionExpr)` is not enough)?

Reply:
I *think* this effectively only checks the return statement is `isImplicit()`. Why we don't treat `{ return foo.<HERE> }` a single expression closure, although the typechecker does?

Reply:
I think this was to match the existing behavior before the solver migration. From memory when the 'isSingleExpressionBody' bool is true the type relation computation avoids penalizing results that don't match the expected type, but still prioritizes ones that do.  I think the reasoning is that the user may intend to write multiple statements, but just hasn't yet.  E.g. if the user intends to write the below
```
someCallToClosureReturningBool {
   print("hello")
   return true
 }
```
and uses code completion at this point:
```
someCallToClosureReturningBool {
   <complete>
 }
```
We should still prioritize results returning bool (in case they really do intend to write a single expression closure), but shouldn't demote "print", for example, as they may add a separate return statement later. If they explicitly wrote a return statement though, like below, we should demote "print":
```
someCallToClosureReturningBool {
   return <complete>
 }
```
I guess it would be more accurate to name it `isImplicitSingleExpressionReturn` or maybe `ignoreExpectedTyMismatch`. Do you think one of those (or something else) would be better?

Reply:
The ignoring invalid types when isSingleExpressionBody is true happens here: https://github.com/apple/swift/blob/main/lib/IDE/CodeCompletion.cpp#L1181

Reply:
Thanks for the explanation! `isImplicitSingleExpressionReturn` looks better I think. Also, could you add some comment?